src/platform/tls_openssl.c

```diff
@@ -0,0 +1,3437 @@
+// with the given SSL connection.
+//
+// @param s Pointer to an @c SSL object.
+//
+// @return Pointer to the application-specific data.
+//
+#define GetSslAuxData(s) BIO_get_app_data(SSL_get_rbio(s))
+
+//
+// @brief Determines the negotiated AEAD and hash algorithms from a TLS session.
+//
+// This function inspects the currently negotiated cipher suite in the given
+// TLS context and maps it to corresponding internal AEAD and hash algorithm
+// types used by the QUIC implementation.
+//
+// Supported ciphers include:
+// - TLS_AES_128_GCM_SHA256
+// - TLS_AES_256_GCM_SHA384
+// - TLS_CHACHA20_POLY1305_SHA256
+//
+// If an unsupported cipher is negotiated, the function asserts.
+//
+// @param[in]  TlsContext Pointer to the TLS context containing the SSL object.
+// @param[out] AeadType   Pointer to receive the negotiated AEAD algorithm type.
+// @param[out] HashType   Pointer to receive the negotiated hash algorithm type.
+//
+void
+CxPlatTlsNegotiatedCiphers(
+    _In_ CXPLAT_TLS* TlsContext,
+    _Out_ CXPLAT_AEAD_TYPE *AeadType,
+    _Out_ CXPLAT_HASH_TYPE *HashType
+    )
+{
+    switch (SSL_CIPHER_get_id(SSL_get_current_cipher(TlsContext->Ssl))) {
+    case 0x03001301U: // TLS_AES_128_GCM_SHA256
+        *AeadType = CXPLAT_AEAD_AES_128_GCM;
+        *HashType = CXPLAT_HASH_SHA256;
+        break;
+    case 0x03001302U: // TLS_AES_256_GCM_SHA384
+        *AeadType = CXPLAT_AEAD_AES_256_GCM;
+        *HashType = CXPLAT_HASH_SHA384;
+        break;
+    case 0x03001303U: // TLS_CHACHA20_POLY1305_SHA256
+        *AeadType = CXPLAT_AEAD_CHACHA20_POLY1305;
+        *HashType = CXPLAT_HASH_SHA256;
+        break;
+    default:
+        CXPLAT_FRE_ASSERT(FALSE);
+    }
+}
+
+//
+// @brief Callback to send TLS handshake data to the QUIC stack.
+//
+// This function is called by OpenSSL to send handshake data over QUIC.
+// It submits the provided buffer to the QUIC connection for transmission.
+// If the submission fails, it sets the TLS error on the QUIC connection.
+//
+// @param[in]  s           Pointer to the SSL connection object.
+// @param[in]  buf         Pointer to the buffer containing data to send.
+// @param[in]  buf_len     Length of the data in @p buf.
+// @param[out] consumed    Number of bytes successfully consumed from @p buf.
+// @param[in]  arg         Unused argument (typically NULL).
+//
+// @return 1 on success, 0 on failure.
+//
+static int QuicTlsSend(SSL *s, const unsigned char *Buf,
+                         size_t BufLen, size_t *Consumed,
+                         void *Arg)
+{
+    CXPLAT_TLS* TlsContext = SSL_get_app_data(s);
+    CXPLAT_TLS_PROCESS_STATE* TlsState = TlsContext->State;
+    struct AUX_DATA *AData = GetSslAuxData(s);
+
+    UNREFERENCED_PARAMETER(Arg);
+
+    //
+    // KeyTypes in msquic map directly to our protection levels
+    //
+    QUIC_PACKET_KEY_TYPE KeyType = (QUIC_PACKET_KEY_TYPE)AData->Level;
+
+    if (TlsContext->ResultFlags & CXPLAT_TLS_RESULT_ERROR) {
+        return -1;
```

Comment:
The comment says this function returns 0 on failure, but all the errors return -1.